source/NVDAObjects/behaviors.py

```diff
@@ -366,6 +367,106 @@ def event_loseFocus(self):
 		super(Terminal, self).event_loseFocus()
 		self.stopMonitoring()
 
+
+class KeyboardHandlerBasedTypedCharSupport(Terminal):
+	"""A Terminal object that also provides typed character support for
+	console applications via keyboardHandler events.
+	This approach is an alternative to monitoring the console output for
+	characters close to the caret, or injecting in-process with NVDAHelper.
+	This class relies on the toUnicodeEx Windows function, and in particular
+	the flag to preserve keyboard state available in Windows 10 1607
+	and later."""
+	#: Whether this object quickly and reliably sends textChange events
+	#: when its contents update.
+	#: Timely and reliable textChange events are required
+	#: to support password suppression.
+	_supportsTextChange = True
+	#: A queue of typed characters, to be dispatched on C{textChange}.
+	#: This queue allows NVDA to suppress typed passwords when needed.
+	_queuedChars = []
+	#: Whether the console got new text lines in its last update.
+	#: Used to determine if typed character/word buffers should be flushed.
+	_hasNewLines = False
+	#: Whether the last typed character is a tab.
+	#: If so, we should temporarily disable filtering as completions may
+	#: be short.
+	_hasTab = False
+
+	def _reportNewText(self, line):
+		# Additional typed character filtering beyond that in LiveText
+		if (
+			not self._hasTab
+			and len(line.strip()) < max(len(speech.curWordChars) + 1, 3)
+		):
+			return
+		if self._hasNewLines:
+			# Clear the typed word buffer for new text lines.
+			speech.clearTypedWordBuffer()
+			self._queuedChars = []
+		super(KeyboardHandlerBasedTypedCharSupport, self)._reportNewText(line)
+
+	def event_typedCharacter(self, ch):
+		if ch == '\t':
+			self._hasTab = True
+			# Clear the typed word buffer for tab completion.
+			speech.clearTypedWordBuffer()
+		else:
+			self._hasTab = False
+		if (
+			(
+				config.conf['keyboard']['speakTypedCharacters']
+				or config.conf['keyboard']['speakTypedWords']
+			)
+			and not config.conf['UIA']['winConsoleSpeakPasswords']
+			and self._supportsTextChange
+		):
+			self._queuedChars.append(ch)
+		else:
+			super(KeyboardHandlerBasedTypedCharSupport, self).event_typedCharacter(ch)
+
+	def event_textChange(self):
+		self._dispatchQueue()
+		super(KeyboardHandlerBasedTypedCharSupport, self).event_textChange()
+
+	@script(gestures=[
+		"kb:enter",
+		"kb:numpadEnter",
+		"kb:tab",
+		"kb:control+c",
+		"kb:control+d",
+		"kb:control+pause"
+	])
+	def script_flush_queuedChars(self, gesture):
+		"""
+		Flushes the typed word buffer and queue of typedCharacter events if present.
+		Since these gestures clear the current word/line, we should flush the
+		queue to avoid erroneously reporting these chars.
+		"""
+		gesture.send()
+		self._queuedChars = []
+		speech.clearTypedWordBuffer()
+
+	def _calculateNewText(self, newLines, oldLines):

```

Comment:
I discussed this briefly with @feerrenrut, but it was a bit difficult to explain my concerns by voice. So, I'm doing it in here.

The flow of things regarding this method is as follows:

1. _calculateNewText is called by the monitor
1. it sets `self._hasNewLines`
1. In `LiveText`, for every line in outLines, `self._reportNewText` is queued
1. `_reportNewText` clears self._queuedChars if self._hasNewLines.

This means that, if 70 new lines are written to the console, NVDA overrides `self._queuedChars` 70 times, which is pretty useless.

One way to fix this is: remove the `if self._hasNewLines:` clause from _reportNewText and move it to _calculateNewText. However, I'm not verry happy with the idea where _calculateNewText does anything more than calculating new text. Therefore, I'd like to know @feerrenrut's opinion about this concern. May be we should keep it as is and revisit this at some point in the future. I could think of several improvements, including using a filter extension point that filters lines before they are spoken. That also avoids doing filtering in _reportNewText, which is not the ideal place to filter text IMO.

Reply:
Moved to `_calculateNewText` for now. This also means we can move `_hasNewLines` to the body of this method.

Reply:
> Moved to _calculateNewText for now 

I think this is fine.